### PR TITLE
feat: add pipeline registry abstraction

### PR DESCRIPTION
**Pull request by aarnphm**
# What does this PR do?

- added `PipelineRegistry` abstraction for better supports for custom pipeline. 
- updates `add_new_pipeline.mdx` (english docs) to reflect the api addition
- migrate `check_task` and `get_supported_tasks` from
  `transformers/pipelines/__init__.py` to
 ` transformers/pipelines/base.py#PipelineRegistry.{check_task,get_supported_tasks}`

Address #17762

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@LysandreJik and @sgugger, would be great if you guys can provide feedback.

### CODE REVIEW

File: src/transformers/pipelines/__init__.py
Comment (sgugger) [inline src/transformers/pipelines/__init__.py:308]:
```suggestion
PipelineRegistry = _PipelineRegistry.from_dict(supported_tasks=SUPPORTED_TASKS, task_aliases=TASK_ALIASES)
```

**Review by sgugger — COMMENTED** [review]
Thanks a lot for your PR! This could be an interesting addition. I'm curious to see what @Narsil and @LysandreJik think of it.

Note that this is only to customize preprocessing and/or postprocessing as this still relies on existing auto-model classes.

File: src/transformers/pipelines/base.py
Comment (aarnphm) [reply src/transformers/pipelines/base.py:1104]: sounds good. I will remove accordingly.

**Comment by aarnphm** [comment]
> Note that this is only to customize preprocessing and/or postprocessing as this still relies on existing auto-model classes.

Should I include inside the docs to use an `AutoModel` class for better readability?

**Comment by aarnphm** [comment]
@Narsil CI failed due to the recent `black` versioning locked?

File: src/transformers/pipelines/__init__.py
Comment (Narsil) [inline src/transformers/pipelines/__init__.py:308]: NIT: As a global and not a class name I would suggest using `PIPELINE_REGISTRY` instead.
Having a registry as a global variable is perfectly fine to me.

Also `from_dict` could be just a regular `__init__` maybe a bit more pythonic ?

**Comment by Narsil** [comment]
Super nice PR thanks for this addition !

Left a few NITs about the structure. Feel free to ignore if you don't agree with them

File: src/transformers/pipelines/__init__.py
Comment (aarnphm) [reply src/transformers/pipelines/__init__.py:308]: I don't have a big preference on this. I will update accordingly.

**Comment by sgugger** [comment]
Yes, you'll need to rebase on main to fix the tests. Failures are due to new releases of black and PyTorch.

**Comment by aarnphm** [comment]
> Yes, you'll need to rebase on main to fix the tests. Failures are due to new releases of black and PyTorch.

understood. Address accordingly

cc @LysandreJik @sgugger @Narsil when you guys have time.

File: src/transformers/pipelines/__init__.py
Comment (aarnphm) [reply src/transformers/pipelines/__init__.py]: Done.

**Review by Narsil — COMMENTED** [review]
LGTM !

Letting a core maintainer approve.

**Review by sgugger — APPROVED** [review]
Good for me! Would like @LysandreJik to also have a look before merging.

**Comment by LysandreJik** [comment]
That's very nice, love this approach! Should make it much much simpler to add custom pipelines.

I only have one request: please add tests :smile:

**Comment by aarnphm** [comment]
cc @LysandreJik for tests. I'm thinking to add a test for log capture output, but I'm not too familiar with transformers logging structure.

**Comment by sgugger** [comment]
If you want to test outputs, we have a util for this called `CaptureStd`. You can see an example of use in [this test](https://github.com/huggingface/transformers/blob/9fe2403bc52e342022ed132561655f84a6b6b7f3/tests/utils/test_cli.py#L30).

**Comment by aarnphm** [comment]
> If you want to test outputs, we have a util for this called `CaptureStd`. You can see an example of use in [this test](https://github.com/huggingface/transformers/blob/9fe2403bc52e342022ed132561655f84a6b6b7f3/tests/utils/test_cli.py#L30).

Thanks. Will update accordingly.

**Comment by aarnphm** [comment]
tests are finished. lmk if any additional testing is required.

File: tests/pipelines/test_pipelines_common.py
Comment (aarnphm) [reply tests/pipelines/test_pipelines_common.py:787]: noted.

**Comment by sgugger** [comment]
Failure is flaky, so merging. Thanks again for your contribution!

**Pull request merged**
